### PR TITLE
fix: throw on impossible state

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

I'm not sure why a partition can get stuck being processed by a session manager but not present in the offset manager

## Changes

But at least let's not stay in a loop forever

A partition could be removed from the offset manager while still processing so...

The question becomes

1) are we not removing session managers correctly on revoke of a partition
2) are we not tracking offsets correctly on assign

(or hopefully not 3) both)

So, explicitly track revoked and re-assigned partitions in the offset manager so we can distinguish between them

## How did you test this code?

developer tests

### CODE REVIEW

File: plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/offset-manager.ts [hunk 1]
```diff
@@ -94,9 +95,12 @@ export class OffsetManager {
 
         if (!inFlightOffsets) {
             gaugeOffsetRemovalImpossible.inc()
-            status.warn('💾', `offset_manager - no inflight offsets found to remove`, { partition })
```
Comment (benjackwhite) [inline plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/offset-manager.ts:97]: Couldn't this happen if the flush callback is triggered after the partition is revoked? Do we handle that case?

**Review by benjackwhite — APPROVED** [review]

File: plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/offset-manager.ts (hunk 1 above)
Comment (pauldambra) [reply plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/offset-manager.ts:97]: OK, changed this around...

The question becomes

1) are we not removing session managers correctly on revoke of a partition
2) are we not tracking offsets correctly on assign

(or hopefully not 3) both)

So, explicitly track revoked and re-assigned partitions in the offset manager so we can distinguish between them

File: plugin-server/src/main/ingestion-queues/session-recording/session-recordings-blob-consumer.ts [hunk 2]
```diff
@@ -280,8 +280,10 @@ export class SessionRecordingBlobIngester {
             if (err.code === CODES.ERRORS.ERR__ASSIGN_PARTITIONS) {
                 /**
                  * The assign_partitions indicates that the consumer group has new assignments.
-                 * We don't need to do anything, but it is useful to log for debugging.
+                 * We explicitly ensire the offset manager is not treating these partitions as revoked
```
Comment (benjackwhite) [inline plugin-server/src/main/ingestion-queues/session-recording/session-recordings-blob-consumer.ts:283]: Finally I'm the one finding typos 😅 

```suggestion
                 * We explicitly ensure the offset manager is not treating these partitions as revoked
```

**Review by benjackwhite — APPROVED** [review]
I find this really hard to follow but it might just be because my brain is fried. 

My gut tells me this is not the right direction... We're adding more and more complexity and state when we probably want less. That said, happy to try it and see if it reveals anything helpful...

File: plugin-server/src/main/ingestion-queues/session-recording/session-recordings-blob-consumer.ts (hunk 2 above)
Comment (pauldambra) [reply plugin-server/src/main/ingestion-queues/session-recording/session-recordings-blob-consumer.ts:283]: No, "ensire" is archaic english for, erm... making something the parent of an animal as an, erm, animal, and...

homer-hedge
